src/waveform.js

```diff
@@ -0,0 +1,189 @@
+import Logger from "./logger";
+import { mousedownCallback } from "./utilities.js";
+
+export default class Waveform {
+  constructor() {
+    this.log = new Logger();
+
+    this.currentTarget;
+    this.canvas;
+    this.canvasDisplayToggle;
+    this.waveformColour;
+    this.waveformOverlayColour;
+
+    this.boundToggleWaveformCanvas = Waveform.toggleWaveformCanvasCallback.bind(
+      this
+    );
+    this.boundMonitorAudioCanPlay = Waveform.monitorAudioCanPlayCallback.bind(
+      this
+    );
+    this.boundMonitorAudioTimeupdate = Waveform.monitorAudioTimeupdateCallback.bind(
+      this
+    );
+  }
+
+  init() {
+    this.canvas = Waveform.createCanvas();
+    this.canvas.addEventListener("click", mousedownCallback);
+
+    this.canvasDisplayToggle = Waveform.createCanvasDisplayToggle();
+    this.canvasDisplayToggle.addEventListener(
+      "change",
+      this.boundToggleWaveformCanvas
+    );
+
+    // this.canvasDisplayToggle.checked = true;
+    // this.canvasDisplayToggle.dispatchEvent(new Event("change")); // defaults on for now
+
+    let bg = document.querySelector("h2.trackTitle");
+    this.waveformColour = window
+      .getComputedStyle(bg, null)
+      .getPropertyValue("color");
+    this.waveformOverlayColour = Waveform.invertColour(this.waveformColour);
+
+    document
+      .querySelector("audio")
+      .addEventListener("canplay", this.boundMonitorAudioCanPlay);
+
+    document
+      .querySelector("audio")
+      .addEventListener("timeupdate", this.boundMonitorAudioTimeupdate);
+  }
+
+  async generateWaveform() {
+    const datapoints = 100;
+    const audio = document.querySelector("audio");
+
+    if (this.currentTarget != audio.src) {
+      this.currentTarget = audio.src;
+
+      this.canvas
+        .getContext("2d")
+        .clearRect(0, 0, this.canvas.width, this.canvas.height);
+
+      const audioContext = new AudioContext();
+      const fs = audioContext.sampleRate;
+      const length = audio.duration;
+      const src = audio.src.split("stream/")[1];
+
+      chrome.runtime.sendMessage(
+        {
+          contentScriptQuery: "renderBuffer",
+          fs: fs,
+          length: length,
+          url: src,
+          datapoints: datapoints
+        },
+        rmsBuffer => {
+          this.log.info("visualizing");
+          let max = rmsBuffer.reduce(function(a, b) {
+            return Math.max(a, b);
+          });
+          for (let i = 0; i < rmsBuffer.length; i++) {
+            let amplitude = rmsBuffer[i] / max;
+            Waveform.fillBar(
+              this.canvas,
+              amplitude,
+              i,
+              datapoints,
+              this.waveformColour
+            );
+          }
+        }
+      );
+    }
+  }
+
+  static toggleWaveformCanvasCallback(event) {
+    this.log.info("waveform toggle: " + event.target.checked);
+    this.canvas.style.display = event.target.checked ? "" : "none";
```

Comment:
Does this work - is `display: ;` a valid rule? If there are browser complaints about invalid styles, `display: inherit;` is a clean way to not change its existing style.

```js
this.canvas.style.display = event.target.checked ? "inherit" : "none";
```